src/subscriptionNotifier/subscriptionNotifier.service.ts

```diff
@@ -0,0 +1,237 @@
+import { Injectable, Logger } from '@nestjs/common';
+import * as cron from 'node-cron';
+import { ConfigService } from '@nestjs/config';
+import { DiscordService } from 'src/discord/discord.service';
+import { SubscriptionService } from 'src/subscription/subscription.service';
+import { Subscriber } from 'prisma/prisma.types';
+import { activities, cabins, subscriptions } from '@prisma/client';
+import { ActivityService } from 'src/activity/activity.service';
+import { CabinService } from 'src/cabin/cabin.service';
+import { EmbedBuilder } from 'discord.js';
+
+type MessageThread = {
+  topic: string | null;
+  mainMessage: string;
+  embeds: EmbedBuilder[];
+};
+
+@Injectable()
+export class SubscriptionNotifierService {
+  private readonly logger = new Logger(SubscriptionNotifierService.name);
+  private task: cron.ScheduledTask;
+
+  constructor(
+    private readonly activityService: ActivityService,
+    private readonly cabinService: CabinService,
+    private readonly configService: ConfigService,
+    private readonly discordService: DiscordService,
+    private readonly subscriptionService: SubscriptionService,
+  ) {
+    if (this.configService.get('NODE_ENV') === 'production') {
+      this.task = cron.schedule('0 16 * * *', this.on4pm);
+      this.logger.log('setup 4pm subscription notifier');
+    }
+  }
+
+  on4pm = () => {
+    this.logger.log('on4pm');
+    this.notifySubscribers()
+      .catch((error) => {
+        this.logger.error(
+          'Sending subscription notifications failed with error.',
+          error,
+        );
+      })
+      .then(() =>
+        this.logger.log('Subscription notifications sent successfully.'),
+      );
+  };
+
+  private async notifySubscribers() {
+    const subscribers = await this.subscriptionService.getSubscribers();
+
+    subscribers?.forEach(async (subscriber) => {
+      const subscriptions = await this.subscriptionService.getSubscriptions(
+        subscriber.subscriberId,
+      );
```

Comment:
To avoid N+1 query here so I am wondering if the grouping should be done in JS. Could use lodash or something to do it?

This would mean we have a single DB query to get all the subscriptions regardless of how many we have. In the future we could do some sort of batching to break it up if needed.

Reply:
Basically you want to reduce how often the db is queried, since this is most likely the bottleneck for performance? I have never worked with lodash, so would have to figure that out first. :)

Reply:
Not so worried about performance as this isn't time critical. More worried about load on the systems I think.

https://lodash.com/docs/#groupBy is the doc for the function I was thinking. Might be an easier way though